org.osgi.test.assertj.monitoring/src/test/java/org/osgi/test/assertj/monitoring/internal/EventRecordingTest.java

```diff
@@ -0,0 +1,134 @@
+package org.osgi.test.assertj.monitoring.internal;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+
+import org.assertj.core.api.ThrowableAssert.ThrowingCallable;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleEvent;
+import org.osgi.framework.BundleListener;
+import org.osgi.framework.FrameworkEvent;
+import org.osgi.framework.FrameworkListener;
+import org.osgi.framework.ServiceEvent;
+import org.osgi.framework.ServiceListener;
+import org.osgi.framework.ServiceReference;
+import org.osgi.test.assertj.monitoring.TimedEvent;
+
+public class EventRecordingTest {
+
+	static List<FrameworkListener> frameworkListeners = new ArrayList<>();
+	static List<ServiceListener> serviceListeners = new ArrayList<>();
+	static List<BundleListener> bundleListeners = new ArrayList<>();
+	static BundleContext bundleContext;
+
+	static Bundle bundle;
+
+	@BeforeAll
+	public static void beforeAll() {
+		bundle = mock(Bundle.class);
+
+		when(bundle.getSymbolicName()).thenReturn("test");
+
+		bundleContext = mock(BundleContext.class);
+
+		doAnswer(invocation -> {
+			Object listener = invocation.getArgument(0);
+			bundleListeners.add((BundleListener) listener);
+			return null;
+		}).when(bundleContext).addBundleListener(Mockito.any(BundleListener.class));
+
+		doAnswer(invocation -> {
+			Object listener = invocation.getArgument(0);
+			bundleListeners.remove(listener);
+			return null;
+		}).when(bundleContext).removeBundleListener(Mockito.any(BundleListener.class));
+
+		doAnswer(invocation -> {
+			Object listener = invocation.getArgument(0);
+			frameworkListeners.add((FrameworkListener) listener);
+			return null;
+		}).when(bundleContext).addFrameworkListener(Mockito.any(FrameworkListener.class));
+
+		doAnswer(invocation -> {
+			Object listener = invocation.getArgument(0);
+			frameworkListeners.remove(listener);
+			return null;
+		}).when(bundleContext).removeFrameworkListener(Mockito.any(FrameworkListener.class));
+
+		doAnswer(invocation -> {
+			Object listener = invocation.getArgument(0);
+			serviceListeners.add((ServiceListener) listener);
+			return null;
+		}).when(bundleContext).addServiceListener(Mockito.any(ServiceListener.class));
+		doAnswer(invocation -> {
+			Object listener = invocation.getArgument(0);
+			serviceListeners.remove(listener);
+			return null;
+		}).when(bundleContext).removeServiceListener(Mockito.any(ServiceListener.class));
+
```

Comment:
When your mock setup gets this complicated, it is much easier, more readable, more type-safe and more maintainable to directly implement the `BundleContext` interface as a concrete class. Especially as you are not using Mockito's `verify()` method anywhere.